<h1>UK ROAD ACCIDENT DATA ANALYSIS</h1>
<h2>INCLUSIVE YEAR 2019 - 2022</h2>
<h3>Analyst: Meloujean Ligon</h3>

<h2>DATA PREPARATION</h2>
<h3>Transfer Data set to folder</h3>

<h2>Importing Libraries</h2>

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from scipy.stats import f_oneway

<h1>DATA FRAME</h1>

In [2]:
accident = pd.read_csv('datasets\\accident_data.csv')

In [3]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [4]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

<h1>Converting Date Field into DATE TIME type</h1>

In [5]:
accident['Accident Date']

0           5/6/2019
1           2/7/2019
2         26-08-2019
3         16-08-2019
4           3/9/2019
             ...    
660674    18-02-2022
660675    21-02-2022
660676    23-02-2022
660677    23-02-2022
660678    28-02-2022
Name: Accident Date, Length: 660679, dtype: object

In [6]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [74]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna('unknown road condition')
accident['Road_Type'] = accident['Road_Type'].fillna('unaccounted')
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('unaccounted')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])

In [9]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [10]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2              unaccounted               Urban            unaccounted   
3       Single carriageway               Urban     Fine no high winds   
4              unaccounted               Urban            unaccounted   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [11]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [12]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')

In [13]:
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')

In [14]:
accident['District Area'] = accident['District Area'].astype('category')

In [15]:
accident['Number_of_Casualties'] = accident['Number_of_Casualties'].astype('category')

In [16]:
accident['Number_of_Vehicles'] = accident['Number_of_Vehicles'].astype('category')

In [17]:
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')

In [18]:
accident['Road_Type'] = accident['Road_Type'].astype('category')

In [19]:
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')

In [20]:
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')

In [21]:
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [22]:
accident['Latitude'] = accident['Latitude'].astype('category')
accident['Longitude'] = accident['Longitude'].astype('category')

In [23]:
accident.dtypes

Index                        object
Accident_Severity          category
Accident Date                object
Latitude                   category
Light_Conditions           category
District Area              category
Longitude                  category
Number_of_Casualties       category
Number_of_Vehicles         category
Road_Surface_Conditions    category
Road_Type                  category
Urban_or_Rural_Area        category
Weather_Conditions         category
Vehicle_Type               category
dtype: object

In [24]:
accident['Accident Date'] = accident['Accident Date'].astype('str')
accident.dtypes
accident['Accident Date'] = accident['Accident Date'].str.strip()
accident['Accident Date'] = accident['Accident Date'].str.replace('/','-')

In [25]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')

In [26]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [27]:
accident.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                         category
Light_Conditions                 category
District Area                    category
Longitude                        category
Number_of_Casualties             category
Number_of_Vehicles               category
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

In [28]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['Day'] = accident['Accident Date'].dt.day
accident['DayOfWeek'] = accident['Accident Date'].dt.dayofweek

In [29]:
latitude_longitude_cor = accident['Latitude'].corr(accident['Longitude'])

In [30]:
latitude_longitude_cor

np.float64(-0.3981137948101014)

In [31]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    2022-02-18  57.374005   
660675  201091NM01881            Slight    2022-02-21  57.232273   
660676  201091NM01935            Slight    2022-02-23  57.585044   
660677  201091NM01964           Serious    2022-02-23  57.214898   
660678  201091NM02142           Serious    2022-02-28  57.575210   

                   Light_Conditions           District Area Longitude  \
0             Darkness - lights lit  Kensington and Chelsea -0.209082   
1                          Daylight  Kensington and Chelsea -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea -0.210215   
3                          Daylight  Kensington and Chelsea -0.202731   
4             Darkness - lights lit  Kensington and Chelsea -0.192487   
...                             ...                     ...       ...   
660674                     Daylight                Highland -3.467828   
660675       Darkness - no lighting                Highland -3.809281   
660676                     Daylight                Highland -3.862727   
660677       Darkness - no lighting                Highland -3.823997   
660678                     Daylight                Highland -3.895673   

       Number_of_Casualties Number_of_Vehicles Road_Surface_Conditions  \
0                         1                  2                     Dry   
1                         1                  2             Wet or damp   
2                         1                  3                     Dry   
3                         1                  4                     Dry   
4                         1                  2                     Dry   
...                     ...                ...                     ...   
660674                    2                  1                     Dry   
660675                    1                  1            Frost or ice   
660676                    1                  3            Frost or ice   
660677                    1                  2             Wet or damp   
660678                    1                  1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2              unaccounted               Urban            unaccounted   
3       Single carriageway               Urban     Fine no high winds   
4              unaccounted               Urban            unaccounted   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  Year  Month  Day  DayOfWeek  
0                                        Car  2019      6    5          2  
1                                        Car  2019      7    2          1  
2                      Taxi/Private hire car  2019      8   26          0  
3       Bus or coach (17 or more pass seats)  2019      8   16          4  
4                              Other vehicle  2019      9    3          1  
...                                      ...   ...    ...  ...        ...  
660674                  

<h4 style="color: blue;">No. 1</h4>

In [75]:
yearly_accidents = accident['Year'].value_counts().sort_index()

In [76]:
yearly_accidents

Year
2019    182115
2020    170591
2021    163554
2022    144419
Name: count, dtype: int64

<H2 style="color: blue;">INSIGHT 1</H2>
<H3>The number of accidents has been decreasing over the years. In 2019, there were 182,115 reported accidents, which dropped to 170,591 in 2020. The decline continued in 2021, with 163,554 accidents, and further decreased to 144,419 in 2022. </H3>

<h4 style="color: blue;">No. 2</h>

In [34]:
road_type_counts = accident['Road_Type'].value_counts()

In [35]:
road_type_counts

Road_Type
Single carriageway    492143
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
unaccounted             4520
Name: count, dtype: int64

<h2 style="color: blue;">INSIGHT 2</h2>
<H3>The most accident-prone road type is "Single carriageway" with 492,143 cases.</H3>

<h4 style="color: blue;">No.3</h4>

In [36]:
month_counts = accident['Month'].value_counts()

In [37]:
month_counts

Month
11    60424
10    59580
7     57445
6     56481
9     56455
5     56352
3     54086
8     53913
1     52872
12    51836
4     51744
2     49491
Name: count, dtype: int64

<h2 style="color: blue;">INSIGHT 3</h2>
<H3>The month with the highest number of accidents is November with 58,687 cases, while the month with the lowest number of accidents is February with 49,042 cases.</H3>

<h4 style="color: blue;">No. 4</h1>

In [38]:
november_accidents = accident[accident['Month'] == 11]
most_common_severity_nov = november_accidents['Accident_Severity'].value_counts().idxmax()

In [39]:
most_common_severity_nov

'Slight'

In [40]:
february_accidents = accident[accident['Month'] == 2]
most_common_severity_feb = february_accidents['Accident_Severity'].value_counts().idxmax()

In [41]:
most_common_severity_feb

'Slight'

<h2 style="color: blue;">INSIGHT 4</h1>
<H3>The majority of accidents in both November and February are categorized as slight</H2>

<h4 style="color: blue;">No. 5</h1>

In [42]:
most_common_road_nov = november_accidents['Road_Surface_Conditions'].value_counts().idxmax()

In [43]:
most_common_road_nov

'Wet or damp'

In [44]:
most_common_road_feb = february_accidents['Road_Surface_Conditions'].value_counts().idxmax()

In [45]:
most_common_road_feb

'Dry'

<h2 style="color: blue;">INSIGHT 5</h1>
<H3>In November, most accidents occurred on wet or damp surfaces, while in February, most accidents occurred on dry surfaces.</H2>

<h4 style="color: blue;">No. 6</h1>

In [46]:
most_common_weather_feb = february_accidents['Weather_Conditions'].value_counts().idxmax()

In [47]:
most_common_weather_feb

'Fine no high winds'

In [48]:
most_common_weather_nov = november_accidents['Weather_Conditions'].value_counts().idxmax()

In [49]:
most_common_weather_nov

'Fine no high winds'

<h2 style="color: blue;">INSIGHT 6</h1>
<H3>The most common weather condition during accidents in both November and February was "Fine, no high winds"</H2>

<h4 style="color: blue;">No. 7</h1>

In [50]:
road_type_counts = accident['Road_Type'].value_counts()

In [51]:
road_type_counts

Road_Type
Single carriageway    492143
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
unaccounted             4520
Name: count, dtype: int64

<h2 style="color: blue;">INSIGHT 7</h1>
<H3>The road type with the highest number of accidents is Single carriageway 492143 with cases.</H2>

<h4 style="color: blue;">No. 8</h1>

In [52]:
fatal_accidents = accident[accident['Accident_Severity'] == 'Fatal']
most_fatal_district = fatal_accidents['District Area'].value_counts().idxmax()

In [53]:
most_fatal_district

'Birmingham'

<h2 style="color: blue;">INSIGHT 8</h1>
<H3>Birmingham has the highest number of fatal accidents</H2>

<h4 style="color: blue;">No. 9</h1>

In [54]:
vehicle_counts = accident['Vehicle_Type'].value_counts()

In [55]:
vehicle_counts

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h2 style="color: blue;">INSIGHT 9</h1>
<h3>Cars are involved in the highest number of accidents, with 497,992 cases, making them the most common vehicle type in accidents.</h2>

<h4 style="color: blue;">No. 10</h1>

In [56]:
accident['Year'].value_counts()

Year
2019    182115
2020    170591
2021    163554
2022    144419
Name: count, dtype: int64

In [57]:
accident_2022 = accident[accident['Year'] == 2022]

In [58]:
accident_2022['Urban_or_Rural_Area'].value_counts().sum()
accident_2022['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          93994
Rural          50425
Unallocated        0
Name: count, dtype: int64

<h2 style="color: blue;">Insight 10</h1>
<h3>In 2022, about 65.8% of accidents in 2022 occurred in urban areas, while 34.9% in rural areas.</h2>

<h4 style="color: blue;">No. 11</h1>

In [59]:
accident_2022['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       115970
Raining no high winds     15195
Other                      4300
Snowing no high winds      2923
unaccounted                2654
Raining + high winds       1140
Fine + high winds          1051
Fog or mist                 902
Snowing + high winds        284
Name: count, dtype: int64

<h2 style="color: blue;">Insight 11</h1>
<h3>About 80% of accidents in 2022 occured in fine conditions with no hign winds</h2>

<h4 style="color: blue;">No. 12</h1>

In [60]:
accident.groupby(['Month', 'Road_Surface_Conditions']).size().unstack()

C:\Users\Meloujean\AppData\Local\Temp\ipykernel_9140\2998902824.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  accident.groupby(['Month', 'Road_Surface_Conditions']).size().unstack()


Road_Surface_Conditions    Dry  Flood over 3cm. deep  Frost or ice  Snow  \
Month                                                                      
1                        22304                   142          4770  1366   
2                        28284                    48          3363  1533   
3                        40312                    48           654   259   
4                        44413                    24            90   109   
5                        46117                    55             8     9   
6                        46983                   109             3     6   
7                        43753                   124             6     8   
8                        41513                    80             5     6   
9                        45110                    69             8     6   
10                       42510                    70           266    48   
11                       28736                   152          1480   504   
12                       17786                    96          7864  2036   

Road_Surface_Conditions  Wet or damp  unknown road condition  
Month                                                         
1                              24224                      66  
2                              16207                      56  
3                              12743                      70  
4                               7061                      47  
5                              10106                      57  
6                               9334                      46  
7                              13495                      59  
8                              12242                      67  
9                              11209                      53  
10                             16619                      67  
11                             29474                      78  
12                             23994                      60

<h2 style="color: blue;">INSIGHT 12</h1>
<h3>From November to January, most of accident occured in wet or damp road surfaces. In contrast February to October the majority of accidents occured in dry surfaces</h2>

<h4 style="color: blue;">No. 13</h1>

In [61]:
accident_year = accident[accident['Year'] == 2019]
accident[accident['Year'] == 2019]['District Area'].mode()[0]

'Birmingham'

In [62]:
accident_year = accident[accident['Year'] == 2020]
accident[accident['Year'] == 2020]['District Area'].mode()[0]

'Birmingham'

In [63]:
accident_year = accident[accident['Year'] == 2021]
accident[accident['Year'] == 2020]['District Area'].mode()[0]

'Birmingham'

In [64]:
accident_year = accident[accident['Year'] == 2022]
accident[accident['Year'] == 2022]['District Area'].mode()[0]

'Birmingham'

<h2 style="color: blue;">INSIGHT 13</h1>
<h3>From 2019 to 2022, Birmingham consistently recorded the highest number of accidents each year.</h2>

<h4 style="color: blue;">No. 14</h1>

In [65]:
birmingham_accidents = accident[accident['District Area'] == 'Birmingham']
severity_birmingham = birmingham_accidents['Accident_Severity'].value_counts().idxmax()

In [66]:
severity_birmingham

'Slight'

<h2 style="color: blue;">INSIGHT NO. 14</h1>
<H3>Most accidents in Birmingham are classified as slight</H2>

<H4 style="color: blue;">No. 15</H1>

In [67]:
birmingham_vehicle_type = birmingham_accidents['Vehicle_Type'].value_counts().idxmax()

In [68]:
birmingham_vehicle_type

'Car'

<H2 style="color: blue;">INSIGHT NO. 15</H1>
<H3>Cars are the most common type of vehicle involved in accidents in Birmingham.</H2>

<h4 style="color: blue;">No. 16</h1>

In [69]:
light_conditions = accident['Light_Conditions'].value_counts()

In [70]:
light_conditions

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h2 style="color: blue;">Insight 16</h1>
<h3>Most accidents happen under "Daylight" conditions with 484880 case</h2>

<h4 style="color: blue;">No. 17, 18, 19, 20</h1>

In [71]:
accident[['Accident Date', 'DayOfWeek']].head(10)

Accident Date  DayOfWeek
0    2019-06-05          2
1    2019-07-02          1
2    2019-08-26          0
3    2019-08-16          4
4    2019-09-03          1
5    2019-09-18          2
6    2019-09-05          3
7    2019-10-03          3
8    2019-10-31          3
9    2019-10-18          4

In [72]:
accident.groupby(['DayOfWeek', 'Accident_Severity']).size().unstack()

C:\Users\Meloujean\AppData\Local\Temp\ipykernel_9140\62436979.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  accident.groupby(['DayOfWeek', 'Accident_Severity']).size().unstack()


Accident_Severity  Fatal  Serious  Slight
DayOfWeek                                
0                   1385    11664   59631
1                   1105    11918   81527
2                   1113    12488   85957
3                   1097    12440   85974
4                   1113    12633   84154
5                   1326    14000   91852
6                   1522    13074   74706

In [73]:
72680, 94550, 99558, 99511, 97900, 107178, 89302

(72680, 94550, 99558, 99511, 97900, 107178, 89302)

<h2 style="color: blue;">INSIGHT 17</h1>
<h3>Saturday (Day 5) has the highest number of accidents, with 91,852 slight and 14,000 serious cases, making it the most accident-prone day of the week.</h2>

<h2 style="color: blue;">INSIGHT 18</h1>
<h3>Sunday (Day 6) records the highest number of fatal accidents (1,522 cases), while Thursdays (Day 3) have the lowest (1,097)</h2> 

<h2 style="color: blue;">INSIGHT 19</h1>
<h3>Monday (Day 0) has the lowest number of total accidents, </h2>

<h2 style="color: blue;">INSIGHT 20</h1>
<h3>Wednesday (99558) and Saturday (107,178) together account for a significant portion of weekly accidents.</h2>